src/server/engine_shard.cc

```diff
@@ -198,6 +202,30 @@ optional<uint32_t> GetPeriodicCycleMs() {
   return clock_cycle_ms;
 }
 
+ssize_t CalculateMemoryBudget(size_t max_memory, size_t used_memory) {
+  if (max_memory >= used_memory) {
+    return max_memory - used_memory;
+  }
+  // negative value indicates memory overuse
+  return -1 * ssize_t(used_memory - max_memory);
```

Comment:
UB because size_t - size_t = size_t

Reply:
```
ssize_t CalculateMemoryBudget(size_t max_memory, size_t used_memory) {
  if (max_memory >= used_memory) {
    return max_memory - used_memory;
  }
  // negative value indicates memory overuse
  return -1 * ssize_t(used_memory - max_memory);
}
```

Reply:
Ok I got, the result is correct because you use overflow of ssize_t type, but it can be UB I'm not sure. please rewrite

Reply:
Do you mean to round size_t to max ssize_t? Because I already implemented this previously
https://github.com/dragonflydb/dragonfly/pull/4142#discussion_r1858484944

Reply:
just update ssize_t CalculateMemoryBudget(ssize_t max_memory, ssize_t used_memory)

Reply:
But we are still converting size_t to ssize_t before calling the function. This is even more dangerous because we did not perform the subtraction.

Reply:
in our case, such converting is absolutely safe, because we will never get such a memory size  in the near future